hack/tools/confdocs/confdocs.go

```diff
@@ -0,0 +1,324 @@
+// Copyright 2021 Zenauth Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+package confdocs
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"golang.org/x/tools/go/packages"
+	"log"
+)
+
+const (
+	unnamedField = "<NONAME>"
+)
+
+type ConfDoc struct {
+	packagesDir      string
+	interfaceName    string
+	interfacePackage string
+	iface            *Interface
+	indexedStructs   map[string]*Struct
+	filteredStructs  map[string]*Struct
+}
+
+func NewConfDoc(packagesDir string, interfaceName string, interfacePath string) *ConfDoc {
+	return &ConfDoc{
+		packagesDir:      packagesDir,
+		interfaceName:    interfaceName,
+		interfacePackage: interfacePath,
+		iface:            nil,
+		indexedStructs:   make(map[string]*Struct),
+		filteredStructs:  make(map[string]*Struct),
+	}
+}
+
+func (cd *ConfDoc) Run() error {
+	fileSet := token.NewFileSet()
+
+	pkgs, err := cd.loadPackages(cd.packagesDir, fileSet)
+
+	if err != nil {
+		return err
+	}
+
+	if err != nil {
+		return err
+	}
+
+	for _, pkg := range pkgs {
+		for _, file := range pkg.Syntax {
+			if pkg.PkgPath == cd.interfacePackage {
+				ifaceRef := cd.findInterface(file, cd.interfaceName)
+				if ifaceRef != nil {
+					cd.iface = ifaceRef
+				}
+			}
+			cd.indexStructs(file, cd.indexedStructs)
+			cd.addMethodsToIndexedStructs(file, cd.indexedStructs)
+		}
+	}
+
+	cd.filterInterfaceImplementingStructs(cd.indexedStructs, cd.iface, cd.filteredStructs)
+
+	return nil
+}
+
+func (cd *ConfDoc) filterInterfaceImplementingStructs(indexedStructs map[string]*Struct, iface *Interface, filteredStructs map[string]*Struct) {
+	for structKey, indexedStruct := range indexedStructs {
+		structImplementsInterface := true
+		for _, interfaceMethod := range iface.Methods {
+			foundInterfaceMethod := false
+
+			for _, method := range indexedStruct.Methods {
+				if interfaceMethod.FunctionName == method.FunctionName {
+					foundInterfaceMethod = true
+				}
+			}
+
+			if !foundInterfaceMethod {
+				structImplementsInterface = false
+				break
+			}
+		}
+		if structImplementsInterface {
+			filteredStructs[structKey] = indexedStruct
+		}
+	}
```

Comment:
I think you can use [`types.Implements`](https://pkg.go.dev/go/types#Implements) for this